File: CPP/No611.cc
/**
 * Created by Xiaozhong on 2020/10/29.
 * Copyright (c) 2020/10/29 Xiaozhong. All rights reserved.
 */
#include <vector>
#include <iostream>
#include <algorithm>

using namespace std;

class Solution {
public:
    int triangleNumber(vector<int> &nums) {
        sort(nums.begin(), nums.end());
        int ans = 0, n = nums.size();
        for (int i = n - 1; i >= 2; --i) {
            int left = 0, right = i - 1;
            while (left < right) {
                if (nums[left] + nums[right] > nums[i]) {
                    ans += right - left;
                    --right;
                }
                else ++left;
            }
        }
        return ans;
    }
};

int main() {
    Solution s;
    vector<int> nums = {2, 2, 3, 4};
    cout << s.triangleNumber(nums) << endl;
}

File: Java/src/No611.java
/*
 * Copyright (c) 8/4/2021, HuXiaozhong. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 */

import java.util.Arrays;

public class No611 {
    public int triangleNumber(int[] nums) {
        Arrays.sort(nums);
        int n = nums.length;
        int ans = 0;
        for (int idx = n - 1; idx >= 2; --idx) {
            int left = 0, right = idx - 1;
            while (left < right) {
                if (nums[left] + nums[right] > nums[idx]) {
                    ans += right - left;
                    right--;
                } else left++;
            }
        }
        return ans;
    }
}
